cargo-frc/src/build.rs

```diff
@@ -0,0 +1,64 @@
+use crate::toolchain::Toolchain;
+use crate::util::str_map;
+use clap::ArgMatches;
+use std::process::Command;
+
+const DEPLOY_TARGET_TRIPLE: &str = "arm-unknown-linux-gnueabi";
+
+pub fn roborio_build(toolchain: Toolchain, bin: Option<&str>, release: bool) -> Result<(), String> {
+    if !toolchain.installed() {
+        return Err(format!(
+            "The {} toolchain is not installed",
+            toolchain.year()
+        ));
+    }
+
+    info!("Building with the {} toolchain", toolchain.year());
+
+    let mut args = vec!["build", "--target", DEPLOY_TARGET_TRIPLE];
+
+    if let Some(bin) = bin {
+        args.push("--bin");
+        args.push(bin);
+    }
+
+    if release {
+        args.push("--release");
+    }
+
+    debug!("Using cargo args {:?}", args);
+
+    let build = Command::new("cargo")
+        .args(args)
+        .env(
+            "RUSTFLAGS",
+            format!(
+                "-C target-cpu=cortex-a9 -C linker={}",
```

Comment:
Shouldn't the linker be passed to cargo via other environment variables rather than rustc directly?

Reply:
I spent a while looking at various rustc and cargo documentation and couldn't find a way to pass the linker through environment variables. The [cargo configuration docs](https://doc.rust-lang.org/cargo/reference/config.html#environment-variables) seem to suggest that all the option in the config file does is pass through to rustc.
> [target.$triple]
> \# This is the linker which is passed to rustc (via `-C linker=`) when the `$triple`
> \# is being compiled for. By default this flag is not passed to the compiler.
> linker = ".."

Reply:
Thinking about this even more, I don't think this is correct at all. Setting `RUSTFLAGS` would break build dependencies… we should set `CARGO_TARGET_..._RUSTFLAGS` or whatever.

It'll be much cleaner to set the linker in cargo's config separately in case it's been set by the user's config as well.

Reply:
I hadn't thought about build dependencies. The cargo env vars aren't super well documented, at least when it comes to `target.$triple`, so it turns out the proper way to configure cargo was to replace the dashes with underscores as well (`CARGO_TARGET_ARM_UNKNOWN_LINUX_GNUEABI_LINKER` and `CARGO_TARGET_ARM_UNKNOWN_LINUX_GNUEABI_RUSTFLAGS`), which was something I hadn't tried when I was figuring this out before.

I'm not convinced about setting the linker in the config file. It just adds complexity, either another command users need to run to generate the config file for a cloned repo or more manual configuration. It also makes it much harder to specify toolchain versions from the cli when building. I think we should be aiming to make this as easy as possible for most teams to use, so the default functionality should be to specify the linker through the env vars, and we should have flags to disable this for more advanced users with custom configs or build systems.

Reply:
I thought cargo also supports CARGO_TARGET_${target}_... with ${target} being the untransformed target name.

> I'm not convinced about setting the linker in the config file.

I'm not suggesting we set the linker in any config files. We need to override any config files.

Reply:
CARGO_TARGET_${target}_... did not work in my testing.

I misinterpreted your last sentence. It sounds like we're in agreement.